### PR TITLE
Integration of 'travis' and `tic` functionality

### PR DESCRIPTION
**Pull request by pat-s**
https://github.com/ropenscilabs/travis/issues/73

Travis: https://github.com/ropenscilabs/travis
tic: https://github.com/ropenscilabs/tic

Travis builds: https://travis-ci.org/pat-s/usethis

### Open tasks

- [ ] Look in sysreqdb for additional dependencies, suggest "addons: apt: ..." entry

- [ ] Write test: How to handle the interactive prompts in tests?
- [x] Export all unexported needed functions from Travis to be able to import them here

### Implementation

New exported functions: `use_ci()`, `use_travis_deploy()`

Main function: `use_ci`:
- Creates YAML templates for `travis` and `appveyor`
- Uses `use_travis_deploy()` to setup deploy access on Travis via ssh
- Adds a Github PAT to the Github repo
- Works interactively so the user can also deny some steps if only partial functionality is needed

Added tic.R templates for different kind of repos under `inst/templates/.

### CODE REVIEW

File: R/ci.R [hunk 1]
```diff
@@ -1,82 +1,104 @@
...
-    ".travis.yml",
-    ignore = TRUE
-  )
+use_ci <- function(path = ".", quiet = FALSE,
+                   services = c("Travis", "Appveyor")) {
+  #' @details
+  #' The following steps will be run:
+  withr::with_dir(path, {
+    #' 1. If necessary, create a GitHub repository via [use_github()]
+    travis:::use_github_interactive() #FIXME: Export function in travis
```
Comment (krlmlr) [inline R/ci.R]: Do we have existing facilities for enabling GitHub in _usethis_?

File: R/ci.R [hunk 2]
```diff
@@ -1,82 +1,104 @@
...
-use_travis <- function(browse = interactive()) {
-  check_uses_github()
-
-  use_template(
-    "travis.yml",
-    ".travis.yml",
-    ignore = TRUE
-  )
+use_ci <- function(path = ".", quiet = FALSE,
+                   services = c("Travis", "Appveyor")) {
```
Comment (krlmlr) [inline R/ci.R]: Using lowercase seems more natural here.

File: DESCRIPTION [hunk 3]
```diff
@@ -47,7 +49,8 @@ Suggests:
     withr
 Remotes: 
     ropensci/git2r,
-    tidyverse/glue
+    tidyverse/glue,
```
Comment (krlmlr) [inline DESCRIPTION]: Can you please apply `use_tidy_description()` ?

File: R/ci.R [hunk 4]
```diff
@@ -1,82 +1,104 @@
...
+#' @importFrom travis travis_set_pat
+#' @param path `[string]`\cr
+#'   The path to the repo to prepare.
+#' @param quiet `[flag]`\cr
+#'   Less verbose output? Default: `FALSE`.
+#' @param services `[character]`\cr
+#'   CI services to add.
 #' @export
 #' @rdname ci
-use_travis <- function(browse = interactive()) {
```
Comment (krlmlr) [inline R/ci.R]: Do we want to keep `use_travis()` and `use_appveyor()` and forward to `use_ci()` (with a message/warning)?

File: R/ci.R [hunk 5]
```diff
@@ -1,82 +1,104 @@
...
+    travis:::use_github_interactive() #FIXME: Export function in travis
+    stopifnot(uses_github())
+
+    if ("Travis" %in% services) {
+      #' 1. Enable Travis via [travis_enable()]
+      travis_enable()
+
+      #' 1. Create a default `.travis.yml` file
+      #'    (overwrite after confirmation in interactive mode only)
+      travis:::use_travis_yml() #FIXME: Export function in travis
```
Comment (krlmlr) [inline R/ci.R]: Can we use a template from _usethis_ ?

File: R/ci.R [hunk 6]
```diff
@@ -1,82 +1,104 @@
...
+    }
 
-  travis_activate(browse)
-  use_travis_badge()
+    repo_type <- travis:::detect_repo_type() #FIXME: Export function in travis
```
Comment (krlmlr) [inline R/ci.R]: What functions do we need to make available in _travis_ for this PR? For some lower-level functions it's better to just copy/move them here.

File: R/ci.R [hunk 7]
```diff
@@ -1,82 +1,104 @@
...
+    #'    via [use_travis_deploy()]
+    if (travis:::needs_deploy(repo_type)) use_travis_deploy() #FIXME: Export function in travis
 
-  use_badge("Travis build status", url, img)
+    #' 1. Create a GitHub PAT and install it on Travis CI via [travis_set_pat()]
+    travis_set_pat()
+  })
 }
 
-travis_activate <- function(browse = interactive()) {
```
Comment (krlmlr) [inline R/ci.R]: Keep, forward to `travis_enable()` and deprecate?

File: R/ci.R [hunk 8]
```diff
@@ -1,82 +1,104 @@
...
 #' @section `use_coverage()`:
 #' Adds test coverage reports to a package that is already using Travis CI.
-#' @rdname ci
+#' @name use_coverage
 #' @param type Which web service to use for test reporting. Currently supports
 #'   [Codecov](https://codecov.io) and [Coveralls](https://coveralls.io).
 #' @export
 use_coverage <- function(type = c("codecov", "coveralls")) {
-  check_uses_travis()
+  travis_is_enabled()
```
Comment (krlmlr) [inline R/ci.R]: This function doesn't have side effects.

The old implementation looked at the presence of a `.travis.yml` file, now we contact the API. I'm not sure which is better. Also, this function is still suggesting to edit `.travis.yml` .

File: R/ci.R [hunk 9]
```diff
@@ -107,7 +129,7 @@ use_coverage <- function(type = c("codecov", "coveralls")) {
 }
 
 use_codecov_badge <- function() {
-  check_uses_github()
+  uses_github()
```
Comment (krlmlr) [inline R/ci.R]: I think `check_uses_github()` does the right thing here?

File: R/ci.R [hunk 10]
```diff
@@ -116,51 +138,72 @@ use_codecov_badge <- function() {
...
-  appveyor_activate(browse)
-  use_appveyor_badge()
+  # generate deploy key pair
+  key <- rsa_keygen()  # TOOD: num bits?
 
-  invisible(TRUE)
-}
+  # encrypt private key using tempkey and iv
+  pub_key <- travis:::get_public_key(key)
+  private_key <- travis:::encode_private_key(key)
```
Comment (krlmlr) [inline R/ci.R]: These two functions should now better live in _tic_ (https://github.com/ropenscilabs/tic/issues/67).

File: R/ci.R [hunk 11]
```diff
@@ -116,51 +138,72 @@ use_codecov_badge <- function() {
...
 
-appveyor_activate <- function(browse = interactive()) {
-  url <- "https://ci.appveyor.com/projects/new"
-  todo("Turn on AppVeyor for this repo at {url}")
-  if (browse) {
-    utils::browseURL(url)
-  }
-}
+  # add to GitHub first, because this can fail because of missing org permissions
+  title <- paste0("travis+tic", if (repo == travis_repo) "" else (paste0(" for ", repo)))
```
Comment (krlmlr) [inline R/ci.R]: I no longer understand the `repo == travis_repo` condition -- I'm aware it's in the original code but we'd better get rid of it?

**Comment by krlmlr** [comment]
@hadley: What's a good time to review/merge this? We're working on getting _travis_ and _tic_ to CRAN by end of summer, but there's a chicken-and-egg problem: _travis_ currently has its own setup code which really belongs in _usethis_.

File: R/ci.R (hunk 7 above)
Comment (pat-s) [reply R/ci.R]: And hence integrate `travis_enable()` from Travis?

File: R/ci.R (hunk 8 above)
Comment (pat-s) [reply R/ci.R]: So which to keep? I thought using all the travis related stuff from `travis` is more tidy than having some "travis_*" functions in 'usethis`.

File: R/ci.R (hunk 9 above)
Comment (pat-s) [reply R/ci.R]: hm, yeah. Idk why I made this change.

File: R/ci.R (hunk 10 above)
Comment (pat-s) [reply R/ci.R]: exported or not?

File: R/ci.R (hunk 4 above)
Comment (pat-s) [reply R/ci.R]: And deprecate? Or remove the `services` arg again from `use_ci()`?

Comment (krlmlr) [reply R/ci.R]: I think _tic_ needs to gain more traction before we can deprecate. At some point `r.rb` from Travis might look for a `tic.R` file and call into _tic_, that would be a good time to start deprecating other approaches.

File: R/ci.R (hunk 8 above)
Comment (krlmlr) [reply R/ci.R]: We still can have a utility wrapper in _usethis_: `check_uses_travis()` could call `travis_is_enabled()` and stop if not.

File: R/ci.R (hunk 10 above)
Comment (krlmlr) [reply R/ci.R]: Exported, the `:::` is caught by the CRAN checks (and effectively these functions are part of _tic_'s API).

Comment (pat-s) [reply R/ci.R]: See respective PRs.

File: R/ci.R (hunk 4 above)
Comment (pat-s) [reply R/ci.R]: Ok, I re-added both and put them into one help file with deprecation warnings.

File: R/ci.R (hunk 8 above)
Comment (pat-s) [reply R/ci.R]: I wrapped `travis_is_enabled()` into `check_uses_travis()`.

File: R/ci.R (hunk 11 above)
Comment (pat-s) [reply R/ci.R]: Seems like an argument to optionally set a different repo to add the key to. But why would anyone ever do this?

Guess we can remove the `travis_repo` arg completely and just use `repo`? I did so now.

**Comment by pat-s** [comment]
@krlmlr addressed all points of the first review

File: R/ci.R [hunk 12]
```diff
@@ -20,6 +78,9 @@ NULL
 #' @export
 #' @rdname ci
 use_travis <- function(browse = interactive()) {
+
+  warning("`use_travis()` is deprecated. Please use `use_ci()` in the future.")
```
Comment (krlmlr) [inline R/ci.R]: A message feels more appropriate here, I'm not sure we're ready to deprecate `use_travis()` .

File: R/ci.R [hunk 13]
```diff
@@ -34,44 +95,78 @@ use_travis <- function(browse = interactive()) {
...
-use_travis_badge <- function() {
+#' @section `use_appveyor()`:
+#' Adds a basic `appveyor.yml` to the top-level directory of a package. This is
+#' a configuration file for the [AppVeyor](https://www.appveyor.com) continuous
+#' integration service for Windows.
+#' @export
+#' @rdname ci
+use_appveyor <- function(browse = interactive()) {
+
+  warning("`use_appveyor()` is deprecated. Please use `use_ci()` in the future.")
```
Comment (krlmlr) [inline R/ci.R]: Same.

File: R/ci.R [hunk 14]
```diff
@@ -124,43 +219,64 @@ use_coveralls_badge <- function() {
...
 }
 
-appveyor_info <- function(base_path = proj_get()) {
-  check_uses_github(base_path)
-  img <- glue(
-    "https://ci.appveyor.com/api/projects/status/github/",
-    "{github_repo_spec()}?branch=master&svg=true"
-  )
-  url <- glue("https://ci.appveyor.com/project/{github_repo_spec()}")
+use_template_tic <- function(..., target = basename(file.path(...))) {
```
Comment (krlmlr) [inline R/ci.R]: This definition seems redundant.

File: R/tidyverse.R [hunk 15]
```diff
@@ -69,14 +71,17 @@ use_tidy_ci <- function(browse = interactive()) {
   )
   use_template("codecov.yml", ignore = TRUE)
 
+  repo_type <- travis:::detect_repo_type()
```
Comment (krlmlr) [inline R/tidyverse.R]: Do we still need `travis:::` here?

File: R/ci.R [hunk 16]
```diff
@@ -1,16 +1,74 @@
...
+      )
+    }
+
+    repo_type <- detect_repo_type()
+
+    if ("appveyor" %in% services) {
+      #' 1. Create a default `appveyor.yml` file
+      #'    (depending on repo type, overwrite after confirmation
+      #'    in interactive mode only)
+      if (travis:::needs_appveyor(repo_type)) travis:::use_appveyor_yml() #FIXME: Export function in travis
```
Comment (krlmlr) [inline R/ci.R]: `needs_appveyor()` can be just copied here.

Do we have an `appveyor.yml` template here?

File: R/ci.R [hunk 17]
```diff
@@ -1,16 +1,74 @@
...
+    }
+
+    #' 1. Create a default `tic.R` file depending on the repo type
+    #'    (package, website, bookdown, ...)
+    use_tic(repo_type)
+
+    #' 1. Enable deployment (if necessary, depending on repo type)
+    #'    via [use_travis_deploy()]
+    if (travis:::needs_deploy(repo_type)) use_travis_deploy() #FIXME: Export function in travis
```
Comment (krlmlr) [inline R/ci.R]: `needs_deploy()` can be copied here.

File: R/tidyverse.R [hunk 18]
```diff
@@ -69,14 +71,17 @@ use_tidy_ci <- function(browse = interactive()) {
...
+  travis:::use_tic_r(repo_type)
```
Comment (krlmlr) [inline R/tidyverse.R]: Do we have a `tic.R` template in this package?

File: R/ci.R [hunk 19]
```diff
@@ -124,43 +219,64 @@ use_coveralls_badge <- function() {
...
-appveyor_activate <- function(browse = interactive()) {
-  url <- "https://ci.appveyor.com/projects/new"
-  todo("Turn on AppVeyor for this repo at {url}")
-  if (browse) {
-    utils::browseURL(url)
-  }
-}
+  # add to GitHub first, because this can fail because of missing org permissions
+  title <- glue("travis+tic for {repo}")
+  github_add_key(pub_key, title = title, info = info)
```
Comment (krlmlr) [inline R/ci.R]: We can just pass `path` here instead of `info`.

File: R/ci.R [hunk 20]
```diff
@@ -124,43 +219,64 @@ use_coveralls_badge <- function() {
...
+#'   The GitHub repo slug, by default obtained through [github_repo()].
+#'   Alternatively, the Travis CI repo ID, e.g. obtained through `travis_repo_id()`.
+#' @param travis_repo `[string]`\cr
+#'   The Travis CI repository to add the private key to, default: `repo`
+#'   (the GitHub repo to which the public deploy key is added).
 #' @export
-#' @rdname ci
-use_appveyor <- function(browse = interactive()) {
-  check_uses_github()
+use_travis_deploy <- function(path = ".", info = travis:::github_info(path),
```
Comment (krlmlr) [inline R/ci.R]: `github_repo()` and `github_repo()` are both exported, not sure we need `info` at all.

File: R/ci.R [hunk 21]
```diff
@@ -124,43 +219,64 @@ use_coveralls_badge <- function() {
...
-  url <- glue("https://ci.appveyor.com/project/{github_repo_spec()}")
+use_template_tic <- function(..., target = basename(file.path(...))) {
+  source <- template_file(...)
+  travis:::safe_filecopy(source, target)
+  message("Added ", target, " from template.")
+  use_build_ignore(target)
+}
 
-  list(url = url, img = img)
+template_file <- function(...) {
```
Comment (krlmlr) [inline R/ci.R]: This is defined twice too.

File: R/helpers.R [hunk 22]
```diff
@@ -97,3 +97,15 @@ view_url <- function(..., open = interactive()) {
   }
   invisible(url)
 }
+
+template_file <- function(...) {
```
Comment (krlmlr) [inline R/helpers.R]: Also, there's `find_template()`, it accepts one argument only but it should be fine.

File: R/ci.R (hunk 14 above)
Comment (pat-s) [reply R/ci.R]: Good catch.

File: R/ci.R (hunk 16 above)
Comment (pat-s) [reply R/ci.R]: Yes, I modified the existing one.

File: R/tidyverse.R (hunk 18 above)
Comment (pat-s) [reply R/tidyverse.R]: Yes, for each repo type under `inst/templates`.

File: R/ci.R (hunk 20 above)
Comment (pat-s) [reply R/ci.R]: I removed `info`.

File: R/ci.R (hunk 21 above)
Comment (pat-s) [reply R/ci.R]: Hm, why did I do that?

File: R/helpers.R (hunk 22 above)
Comment (pat-s) [reply R/helpers.R]: So much similar named functions with the same functionality.

**Comment by pat-s** [comment]
> Thanks. I think we no longer really need travis::: anywhere, can you confirm?

Confirmed. Got rid of all.

File: NAMESPACE [hunk 23]
```diff
@@ -82,10 +83,20 @@ export(use_tidy_support)
...
 export(use_travis)
+export(use_travis_deploy)
 export(use_usethis)
 export(use_version)
 export(use_vignette)
 export(write_over)
 export(write_union)
 import(fs)
 importFrom(glue,glue)
+importFrom(openssl,rsa_keygen)
```
Comment (krlmlr) [inline NAMESPACE]: The `importFrom()` calls here contradict the coding style of the rest of the package.

**Comment by hadley** [comment]
This is a very big PR without any prior discussion. At a minimum, we do not want to use tic for usethis, as we want to keep our build process as consistent as possible across r-lib/tidyverse (many of us maintain multiple packages so having different processes in different packages is unduly burdensome).

Can you start with an issue giving an overview of the proposed changes?

**Comment by pat-s** [comment]
Hi Hadley,

I did not know that there was no prior discussion between @krlmlr and your team. 
I thought both parties agreed on this.
I was just responsible to implement the changes here so we can bring `tic` to CRAN with a proper (usethis) setup.

> so having different processes in different packages is unduly burdensome

This is at least one point for which `tic` could help.
In any case, I leave the rest of the possible integration discussion to @krlmlr :)

File: NAMESPACE (hunk 23 above)
Comment (pat-s) [reply NAMESPACE]: Yes, was still an importFrom leftover. Cleared now.

**Comment by krlmlr** [comment]
We discussed with @karthik and agreed that the setup code for _tic_ that currently lives in _travis_ is better suited in _usethis_. This is what the PR is about.

I'll take care of the issue that describes the proposed changes.

Thanks @pat-s for your help.

**Comment by pat-s** [comment]
@krlmlr Just came cross this https://github.com/krlmlr/r-appveyor/issues/108 which gave me headaches for hours. Should be the default in the appveyor template (as long as it applies). Otherwise users will be highly confused why tests are skipped.

**Comment by hadley** [comment]
Closing in favour of the more specific PRs

**Pull request closed without merging**